### PR TITLE
Allow multiple fee policies

### LINKED ISSUE #2495: feat: Add support for multiple fee policies
### Problem
`Error::MultipleFeePolicies` should be removed by implementing a support for multiple fee policies:

https://github.com/cowprotocol/services/blob/main/crates/driver/src/domain/competition/solution/fee.rs#L79
https://github.com/cowprotocol/services/blob/main/crates/driver/src/domain/competition/solution/scoring.rs#L147

### PR DESCRIPTION
**Pull request by Mateo-mro**
# Description
Support for multiple fee policies

# Changes
- Allow multiple fee policies in autopilot and driver
- Apply fee policies one by one, where each applied policy creates a new solution as a base to apply next fee policy

## How to test
1. e2e test
2. unit tests

## Related Issues

Fixes #2495

### CODE REVIEW

**Comment by squadgazzz** [comment]
Is it requested to be reviewed? Otherwise, may I ask you to convert it to a draft to avoid receiving notifications on the WIP PR?

File: crates/driver/src/domain/competition/solution/fee.rs [hunk 1]
```diff
@@ -75,46 +75,70 @@ impl Fulfillment {
 
     /// Computed protocol fee in surplus token.
     fn protocol_fee(&self, prices: ClearingPrices) -> Result<eth::TokenAmount, Error> {
-        // TODO: support multiple fee policies
-        if self.order().protocol_fees.len() > 1 {
-            return Err(Error::MultipleFeePolicies);
-        }
+        let mut current_order_price_limits = PriceLimits {
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/fee.rs:78]: Where is this actually used/read?

File: crates/driver/src/domain/eth/mod.rs [hunk 2]
```diff
@@ -195,6 +195,12 @@ impl TokenAmount {
                 .into(),
         )
     }
+
+    pub fn checked_sub(self, other: TokenAmount) -> Option<TokenAmount> {
```
Comment (sunce86) [inline crates/driver/src/domain/eth/mod.rs:199]: why not reusing num::CheckedSub

**Review by sunce86 — COMMENTED** [review]
After some thinking, can we maybe start this implementation from writing driver tests in a TDD manner (tests that will initially fail obviously)?

For example, one example test (that would happen in practice) is to have two policies, one price improvement and one volume based. And with that test, we can better debug the changes.

File: crates/driver/src/domain/eth/mod.rs (hunk 2 above)
Comment (Mateo-mro) [reply crates/driver/src/domain/eth/mod.rs:199]: Mostly to avoid doing `value_a.0.checked_sub(value_b.0)`, and having `value_a.checked_sub(value_b)`.

Comment (MartinquaXD) [reply crates/driver/src/domain/eth/mod.rs:199]: You could then also implement `num::CheckSub` on the wrapper type, no?

Comment (Mateo-mro) [reply crates/driver/src/domain/eth/mod.rs:199]: @MartinquaXD what do you mean? it is implemented in the wrapper type `TokenAmount` :thinking:

File: crates/driver/src/domain/competition/solution/fee.rs [hunk 3]
```diff
@@ -75,46 +75,83 @@ impl Fulfillment {
...
+            sell: self.order().sell.amount,
+            buy: self.order().buy.amount,
+        };
+        let mut fee = eth::TokenAmount::default();
+        self.order()
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:83]: style nit: Since this iterator is not very functional anyway (compute result outside the loop) having a simple `for` loop would help with readability here.

File: crates/driver/src/domain/competition/solution/fee.rs [hunk 4]
```diff
@@ -75,46 +75,83 @@ impl Fulfillment {
...
-                        buy_amount: quote.buy.amount.0,
-                        fee_amount: quote.fee.amount.0,
-                    },
-                )?;
-                self.calculate_fee(price_limits, prices, *factor, *max_volume_factor)
-            }
-            Some(FeePolicy::Volume { factor }) => self.fee_from_volume(prices, *factor),
-            None => Ok(0.into()),
-        }
+                // Recalculate the current price limits in order to apply the next protocol fees
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:119]: So we actually have to match on the `side` here?
Shouldn't it be irrelevant whether we increase the sell token price or decrease the buy token price for any given order?

File: crates/autopilot/src/domain/fee/mod.rs [hunk 5]
```diff
@@ -122,26 +122,24 @@ impl ProtocolFees {
...
-                        }
-                    })
-                    .and_then(|policy| match policy {
-                        policy::Policy::Surplus(variant) => variant.apply(&order),
-                        policy::Policy::PriceImprovement(variant) => variant.apply(&order, quote),
-                        policy::Policy::Volume(variant) => variant.apply(&order),
-                    })
-                    .into_iter()
-                    .collect_vec()
+            self.fee_policies
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/fee/mod.rs:125]: There is currently still an early return for the partner fee. AFAIU supporting multiple fees is specifically intended to allow setting a partner fee alongside a regular protocol fee.

File: crates/driver/src/domain/eth/mod.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/driver/src/domain/eth/mod.rs:199]: Yes, but Dusan was referring to implementing it like:
```rust
impl num::CheckedSub for TokenAmount {
//
}
```
Instead of having it in the regular `impl TokenAmount` block to possibly get some nicer ergonomics elsewhere.

Comment (Mateo-mro) [reply crates/driver/src/domain/eth/mod.rs:199]: Ah, I see what you mean :+1: 

> Also the checked_sub function on TokenAmount should check that the tokens are actually the same. Otherwise doing math with them doesn't make sense.
but as it is right now `TokenAmount`, what you mention is responsibility of the caller. `TokenAmount` could have another field for the `address` too.

File: crates/autopilot/src/domain/fee/mod.rs (hunk 5 above)
Comment (Mateo-mro) [reply crates/autopilot/src/domain/fee/mod.rs:125]: Aha, ok, I thought partner fee was still "exclusive", got it! thanks!

File: crates/driver/src/domain/eth/mod.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/driver/src/domain/eth/mod.rs:199]: Ah, sorry. Got confused. I though the `TokenAmount` also contains the token address. In that case indeed no additional check is required. 👌

File: crates/driver/src/tests/setup/solver.rs [hunk 6]
```diff
@@ -62,7 +62,7 @@ impl Solver {
                 order::Side::Buy if config.quote => {
                     "22300745198530623141535718272648361505980416".to_owned()
                 }
-                order::Side::Buy => match quote.order.fee_policy {
+                order::Side::Buy => match quote.order.fee_policy.first().unwrap() {
```
[the remark this answers is not in the document]
Comment (Mateo-mro) [reply crates/driver/src/tests/setup/solver.rs:65]: Actually we could do this as a follow up, since right now we do not necessary need tests which checks the multiple fee calculation on the solver side :thinking: what do you guys think?

File: crates/autopilot/src/domain/fee/mod.rs (hunk 5 above)
Comment (MartinquaXD) [reply crates/autopilot/src/domain/fee/mod.rs:123]: style nit: by now it probably makes sense to have a separate early return for the exempt case right after computing the protocol fee. That way you get rid of the indentation of the `else` branch and can move the `quote_` stuff closer to where it's needed.
```
let partner_fee = //...
if exempt {
    return partner_fee;
}
let protocol_fees = //...
protocol_fees.into_iter().chain(partner_fee).collect()
```

File: crates/driver/src/domain/competition/solution/fee.rs [hunk 7]
```diff
@@ -75,46 +76,53 @@ impl Fulfillment {
...
-                )?;
-                self.calculate_fee(price_limits, prices, *factor, *max_volume_factor)
-            }
-            Some(FeePolicy::Volume { factor }) => self.fee_from_volume(prices, *factor),
-            None => Ok(0.into()),
+            // Recalculate the current price limits in order to apply the next protocol fees
+            // to the new amount
+            current_order_price_limits.buy = current_order_price_limits
+                .buy
+                .checked_add(&current_fee)
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/fee.rs:121]: You need to check in what token the fee is denominated (either sell token or weth). If it's in sell token it would be easier to decrease the sell amount and if it's in weth it's probably equally much work to adjust sell or buy token.

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 8]
```diff
@@ -194,10 +190,16 @@ impl Trade {
...
         };
 
-        let protocol_fee = self.policies.first().map(protocol_fee).transpose();
+        let amount = self
+            .policies
+            .iter()
+            .try_fold(TokenAmount::default(), |acc, protocol_fee| {
+                acc.checked_add(&calc_protocol_fee(protocol_fee).ok()?)
+            })
+            .unwrap_or_default();
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/scoring.rs:199]: This seems slightly sketchy (was already before). To get better introspection here we should log a warning when we actually fallback to 0 due to some error.

File: crates/driver/src/domain/eth/mod.rs [hunk 9]
```diff
@@ -197,6 +200,54 @@ impl TokenAmount {
...
+    }
+}
+
+impl num::CheckedAdd for TokenAmount {
+    fn checked_add(&self, other: &Self) -> Option<Self> {
+        self.0.checked_add(other.0).map(Into::into)
+    }
+}
+
+impl Div<Self> for TokenAmount {
```
Comment (MartinquaXD) [inline crates/driver/src/domain/eth/mod.rs:237]: Making it easier to create div by zero errors does not seem like a good idea. All the other unchecked operations are way less likely to happen (actually unchecked sub is also pretty risky).

Comment (Mateo-mro) [reply crates/driver/src/domain/eth/mod.rs:237]: @MartinquaXD This `trait` is mandatory to be implemented for the `checked_div` trait. Because one is the super trait of the other. All the types which implements `checked_div` *must* implement `div` too.

File: crates/driver/src/domain/competition/solution/fee.rs (hunk 7 above)
Comment (Mateo-mro) [reply crates/driver/src/domain/competition/solution/fee.rs:121]: I wasn't 100 % sure about it. Good point! I still need to extend it with tests which do `Sell` operation instead of `Buy, so a good place to test this! Thank you!

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 10]
```diff
@@ -111,6 +111,89 @@ async fn protocol_fee_test_case(test_case: TestCase) {
...
+            solver: Amounts {
+                sell: 30.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 45.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+        },
+        expected_score: 15.ether().into_wei(),
```
Comment (sunce86) [inline crates/driver/src/tests/cases/protocol_fees.rs:149]: This should be `20` instead of `15`, which would indicate some issue in the `scoring` logic.

File: crates/driver/src/domain/competition/solution/fee.rs [hunk 11]
```diff
@@ -75,46 +76,63 @@ impl Fulfillment {
...
+                        .buy
+                        .checked_add(&current_fee)
+                        .ok_or(Math::Overflow)?;
+                }
+                Side::Sell => {
+                    current_order_price_limits.sell = current_order_price_limits
+                        .sell
+                        .checked_sub(&current_fee)
+                        .ok_or(Math::Negative)?;
+                }
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/fee.rs:131]: Why recalculating the price limits? Why not calculating the protocol fee for first policy, then adjusting `executed` and `fee` of the `Fulfilment`, and then calculating next fee policy etc? 

I find it very hard to understand the logic behind altering the limit prices, even if it works somehow (but I might be biased since I already know this code well).

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 12]
```diff
@@ -194,10 +190,16 @@ impl Trade {
             order::FeePolicy::Volume { factor } => Ok(self.volume_fee(*factor)?.amount),
         };
 
-        let protocol_fee = self.policies.first().map(protocol_fee).transpose();
+        let amount = self
+            .policies
+            .iter()
+            .try_fold(TokenAmount::default(), |acc, protocol_fee| {
+                acc.checked_add(&calc_protocol_fee(protocol_fee).ok()?)
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/scoring.rs:197]: I don't think it won't work this way. After each calculated protocol fee, `custom_prices` change, and `custom_prices` depend on the `executed` and `fee` fields of the fulfilment, which again change after each applied protocol fee.

**Review by sunce86 — COMMENTED** [review]
Good start. I would recommend trying out a different logic to actually change `executed` and `fee` fields of the `Fulfilment` and call the protocol fee function multiple times. This might require some refactoring of Fulfilment struct or even creating another struct to hold intermediary states as protocol fees are applied one by one.

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 13]
```diff
@@ -158,17 +157,50 @@ impl Trade {
...
-        // TODO: support multiple fee policies
-        if self.policies.len() > 1 {
-            return Err(Error::MultipleFeePolicies);
+    fn protocol_fees(&self) -> Result<eth::Asset, Error> {
+        let mut amount = TokenAmount::default();
+        let mut current_trade = self.clone();
+        let mut trade_fee: TokenAmount = self.fulfillment.fee().0.into();
+        for (i, protocol_fee) in self.fulfillment.order().protocol_fees.iter().enumerate() {
+            let fee = current_trade.protocol_fee(protocol_fee)?;
+            if i < self.fulfillment.order().protocol_fees.iter().len() - 1 {
```
Comment (Mateo-mro) [inline crates/driver/src/domain/competition/solution/scoring.rs:169]: This is temporary, it is dirty. Need to make it better

File: crates/driver/src/domain/eth/mod.rs (hunk 9 above)
Comment (MartinquaXD) [reply crates/driver/src/domain/eth/mod.rs:237]: Interesting. I wasn't aware of that. Wouldn't it make more sense the other way around? 🤔 
In any case then I'd say this is fine.

Comment (Mateo-mro) [reply crates/driver/src/domain/eth/mod.rs:237]: > Wouldn't it make more sense the other way around? 🤔
@MartinquaXD No, the reason is because `checked_div()` under the hood is something like:
```rust
pub fn checked_div(&self, other: &Self) -> Option<Self> {
   if other.is_zero() {
      return None;
   }
   Some(self.div(other))
}
 ```
 (it is dirty code just for explanation).
 So it uses `div`, that's the reason it is forcing its implementation.

Comment (MartinquaXD) [reply crates/driver/src/domain/eth/mod.rs:237]: Sure, but there are 2 ways to skin this cat. You could also do it the other way around which would conceptually make more sense to me. Specifically because of the case we are currently talking about:
```rust
fn div(&self, other: &Self) -> Self {
    self.checked_div(other).unwrap()
}
```
But I guess it was done the way it was done because `Div` is already built-in for the primitive types in the standard library so `CheckedDiv` can easily extend those. Also you might not want to rely on the compiler optimizing the unnecessary checks away when doing `.unwrap()`.
If you don't think about how things are implemented under the hood and only think about which type should extend which other type it makes more sense to me to have it the other way around. In fact this is also what they did in the standard library for `Ord` and `PartialOrd`.
`Ord` (which neglects edge cases) extends `PartialOrd` (which is aware of these edge cases).

Comment (Mateo-mro) [reply crates/driver/src/domain/eth/mod.rs:237]: I totally understand your point of view and what you mean. But the guys who wrote those traits in Rust did it like that (I believe, I may be wrong) because if you do this:

```rust
fn div(&self, other: &Self) -> Self {
    self.checked_div(other).unwrap()
}
```
Then if it panics you miss the reason because of the `unwrap()`, whereas the other way around (how it is really implemented), the code fails because of a real panic ([with a message](https://doc.rust-lang.org/std/ops/trait.Div.html)). You can argue that you can do `.expect()` with a custom message, but then you lose the granularity, because the code which is called may panic for different reasons. And if you want to keep the original panic message then you need to add boilerplate code which can be avoided by not overriding the panic.

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 14]
```diff
@@ -132,6 +137,36 @@ impl Solution {
         self.gas
     }
 
+    pub fn calculate_custom_prices(
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/mod.rs:140]: This can be private, right?

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 15]
```diff
@@ -158,17 +157,85 @@ impl Trade {
...
+            .iter()
+            .rev()
+            .enumerate()
+        {
+            let fee = current_trade.protocol_fee(protocol_fee)?;
+            amount = amount
+                .checked_add(&fee)
+                .ok_or(Error::Math(Math::Overflow))?;
+            // Do not apply the last fee since it won't be used again, it minimizes the
+            // changes of negative trade fee
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/scoring.rs:179]: Can you give a bit more context on why we don't use one fee here?
Also you can make this slightly nicer to read by doing `iter.().enumerate().rev()`. Then you can do `if i==0`.
Also would this exception cause problems if there is only 1 fee?

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 16]
```diff
@@ -158,17 +157,85 @@ impl Trade {
...
+            },
+            buy: match side {
+                Side::Sell => executed.0 + fulfillment_fee.0,
+                Side::Buy => {
+                    (executed.0)
+                        .checked_mul(uniform_prices.buy)
+                        .ok_or(Math::Overflow)?
+                        .checked_div(uniform_prices.sell)
+                        .ok_or(Math::DivisionByZero)?
+                        + fulfillment_fee.0
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/scoring.rs:226]: might as well use `checked_sub` and `checked_add` everywhere.

File: crates/driver/src/domain/competition/solution/scoring.rs (hunk 15 above)
Comment (Mateo-mro) [reply crates/driver/src/domain/competition/solution/scoring.rs:179]: > Then you can do if i==0.
Also would this exception cause problems if there is only 1 fee?

Not at all. It is not an exemption, and actually we can even remove it. The reason it is there is because when you calculate the last fee, it is not needed to recalculate again the `custom_prices`, because we won't use it anymore for the next fee calculation (since it was the last one), so I added that there for two reasons:
- avoid extra operation
- avoid potentially triggering an `Overflow` since we are doing operations which aren't meant to be done

File: crates/driver/src/domain/competition/solution/mod.rs [hunk 17]
```diff
@@ -152,36 +187,17 @@ impl Solution {
...
-                            .checked_div(uniform_prices.sell)
-                            .ok_or(error::Math::DivisionByZero)?
-                            + trade.fee().0
-                    }
-                },
-            };
-            trades.push(scoring::Trade::new(
-                trade.order().sell,
-                trade.order().buy,
+            let custom_prices = Self::calculate_custom_prices(
```
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/mod.rs:190]: nit: not sure but maybe this method should even be exposed on `Trade` iteself?

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 18]
```diff
@@ -111,6 +111,320 @@ async fn protocol_fee_test_case(test_case: TestCase) {
...
+        expected_score: 20.ether().into_wei(),
+        fee_handler: FeeHandler::Driver,
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn double_surplus_protocol_fee_sell_order_not_capped() {
```
Comment (fleupold) [inline crates/driver/src/tests/cases/protocol_fees.rs:207]: nit: I like the extended tests, but am afraid they might become a bit too large to maintain. Is there any value in having the double and triple test?

File: crates/driver/src/tests/setup/solver.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (fleupold) [reply crates/driver/src/tests/setup/solver.rs:65]: If not too much work I'd do it now (or at least assert `quote.order.fee_policy.len() == 1` so that once a future developer crosses that bridge they fail explicitly in the place that needs a change.

**Review by fleupold — COMMENTED** [review]
Looks already very good to me. Could we also add an e2e test that checks partner fee + price improvement (or surplus) fee. This is the likely scenario we want to support asap.

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 18 above)
Comment (Mateo-mro) [reply crates/driver/src/tests/cases/protocol_fees.rs:207]: @fleupold the double test and triple tests were really good for me to debug and catch the errors, since they are symmetrical. But they can be removed as they don't bring any extra value.

**Comment by Mateo-mro** [comment]
> Looks already very good to me. Could we also add an e2e test that checks partner fee + price improvement (or surplus) fee. This is the likely scenario we want to support asap.

@fleupold From the driver perspective, a partner fee is just a volume fee, there is no "partner fee". The autopilot sends the partner fee to the driver as volume fee. So, in order to test it, in the e2e test `combined_protocol_fees`, I made the partner fee order as a limit order in order for it to apply the 2 fees (surplus + partner fee).

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 18 above)
Comment (Mateo-mro) [reply crates/driver/src/tests/cases/protocol_fees.rs:207]: I will leave the tests in the PR until @sunce86 reviews this PR, and then I will remove them if that is ok with you.

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 19]
```diff
@@ -158,17 +158,84 @@ impl Trade {
...
+            buy: match side {
+                Side::Sell => executed.0 + fulfillment_fee.0,
+                Side::Buy => (executed.0)
+                    .checked_mul(uniform_prices.buy)
+                    .ok_or(Math::Overflow)?
+                    .checked_div(uniform_prices.sell)
+                    .ok_or(Math::DivisionByZero)?
+                    .checked_add(fulfillment_fee.0)
+                    .ok_or(Math::Overflow)?
+                    .checked_sub(current_protocol_fee.0)
```
Comment (squadgazzz) [inline crates/driver/src/domain/competition/solution/scoring.rs:227]: Why do we need to return an error here instead of using `saturating_sub`?

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 20]
```diff
@@ -61,30 +67,24 @@ impl Scoring {
...
 }
 
 impl Trade {
     pub fn new(
-        sell: eth::Asset,
-        buy: eth::Asset,
-        side: Side,
+        fulfillment: &Fulfillment,
         executed: order::TargetAmount,
+        uniform_prices: ClearingPrices,
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/scoring.rs:80]: Let's check if it is possible to calculate the same thing without using the `Fulfillment` and uniform clearing prices. This is very important to achieve because it proves that the same calculation can be done on an onchain observable settlement (where `fulfillment.executed` and `fulfillment.fee` are not visible and also where uniform clearing prices also potentially won't be visible soon).

So, if my math is correct, we should be able to:
1. Calculate the protocol fee based on the `Trade` data as it was before this PR
2. Create new Trade with the same data as old Trade but with adjusted custom prices to exclude protocol fee =>  new_custom_prices
3. Calculate next protocol fee using newly created Trade

How do we get from original custom prices to new custom prices while working only with protocol fee:

```
let custom_price = match trade.order().side {
                order::Side::Sell => scoring::CustomClearingPrices {
                    sell: custom_price.sell + protocol_fee,
                    buy: custom_price.buy,
                },
                order::Side::Buy => scoring::CustomClearingPrices {
                    sell: custom_price.sell,
                    buy: custom_price.buy - protocol_fee,
                },
            };
```
(don't ask me how I came to this) 😄 The math is just a bit more complex if driver doesn't use the original trade values to build the custom prices (for example if scales them), but we can tackle that later.

Can we try this out and see if it works?

**Review by sunce86 — COMMENTED** [review]
Almost there. Fulfillment logic looks good! Scoring logic (i think) could and should be improved to work only with onchain visible data. If we are not able to achieve this, than there is no point in implementing this feature in this way since accounting wont be able to recover the protocol fees.

File: crates/driver/src/domain/competition/solution/scoring.rs (hunk 20 above)
Comment (Mateo-mro) [reply crates/driver/src/domain/competition/solution/scoring.rs]: Thanks @sunce86 , it works amazingly!!

File: crates/driver/src/tests/cases/protocol_fees.rs [hunk 21]
```diff
@@ -111,6 +111,320 @@ async fn protocol_fee_test_case(test_case: TestCase) {
...
+            // -> First fee policy:
+            // Receive 10 ETH more than quoted, half of which gets captured by the protocol (10 ETH)
+            // Fee = 10 ETH * 0.5 => 5 ETH
+            // -> Second fee policy:
+            // New buy amount in the Order: 50 ETH + 5 ETH (fee) = 55 ETH
+            // New surplus: 5 ETH, fee 2.5 ETH
+            // Total fee: 7.5 ETH
+            solver: Amounts {
+                sell: 20.ether().into_wei(),
+                buy: 30.ether().into_wei(),
```
Comment (MartinquaXD) [inline crates/driver/src/tests/cases/protocol_fees.rs:366]: Is this a test for a partially fillable order? Otherwise I'm not sure how we arrive at these values.
If it is indeed a partially fillable test it would be good to mention it somewhere. (same goes for all partially fillable tests)

**Review by MartinquaXD — APPROVED** [review]
Code looks alright to me and I could follow most of the tests. 👍 
Would be good to get some more clarification on one of the tests. Also an approval from @sunce86 as master of fees would be great.

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 21 above)
Comment (Mateo-mro) [reply crates/driver/src/tests/cases/protocol_fees.rs:366]: @MartinquaXD of course! I changed the numbering and the wording to make it clearer. Please, let me know if it is ok like this :pray:

**Review by MartinquaXD — APPROVED** [review]

**Comment by sunce86** [comment]
Related to https://github.com/cowprotocol/services/pull/2595#discussion_r1561332809, I've adjusted the way we build custom prices for the next fee calculation iteration: https://github.com/cowprotocol/services/pull/2595/commits/7c00d6d9fd0f9070ec2e61d7ec240dcb3cd0f71b

Note that this is not actually needed for the driver itself (since the driver uses traded amounts directly to build custom prices). But, if external solver were to not use reference driver, it could build custom prices that are scaled with some factor. For example, if traded amounts are sell=1000, buy = 2000, then clearing prices are sell=2000,buy=1000, but they could also be sell=2,buy=1 because the exchange rate is what is important. In this case, math would not work without this fix, where we use `trade.executed` and old custom prices to rebuild traded amounts and build new prices.

Bottom line, I think this code now is ready to be replicated to autopilot and also to solver accounting as is, without any adjustments. 🎉

File: crates/autopilot/src/domain/fee/mod.rs [hunk 22]
```diff
@@ -116,33 +121,27 @@ impl ProtocolFees {
...
+                    (false, OrderClass::Market) => Some(&fee_policy.policy),
+                    _ => None,
+                }
+            })
+            .flat_map(|policy| match policy {
+                policy::Policy::Surplus(variant) => variant.apply(&order),
+                policy::Policy::PriceImprovement(variant) => variant.apply(&order, quote),
+                policy::Policy::Volume(variant) => variant.apply(&order),
+            })
+            .chain(partner_fee)
```
Comment (sunce86) [inline crates/autopilot/src/domain/fee/mod.rs:143]: After all, I am not sure if we should give partner fee a priority here or not. Let's keep it this way for now, although we might revaluate in the future.

File: crates/driver/src/domain/competition/solution/scoring.rs [hunk 23]
```diff
@@ -158,17 +162,74 @@ impl Trade {
...
+                    .ok_or(Math::Negative)?,
+            },
+        })
+    }
+
+    /// Protocol fee is defined by a fee policy attached to the order.
+    ///
+    /// Denominated in SURPLUS token
+    fn protocol_fee(&self, fee_policy: &FeePolicy) -> Result<TokenAmount, Error> {
+        let calc_protocol_fee = |policy: &FeePolicy| match policy {
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/scoring.rs:231]: nit: I think this closure is redundant now, we can just use the inner code

File: crates/driver/src/domain/competition/solution/trade.rs [hunk 24]
```diff
@@ -76,10 +80,43 @@ impl Fulfillment {
         }
     }
 
+    pub fn calculate_custom_prices(
```
Comment (sunce86) [inline crates/driver/src/domain/competition/solution/trade.rs:83]: nit: maybe add a fn comment, where it should be explained that here custom prices are built using the uniform clearing prices and the fee, so that custom prices represent the actual traded amounts as seen from the user perspective (amounts flowing in and out of their wallet)

File: crates/driver/src/tests/cases/protocol_fees.rs (hunk 18 above)
Comment (sunce86) [reply crates/driver/src/tests/cases/protocol_fees.rs:207]: Yes, we can probably remove double test if we already have triple test that covers it.

**Review by sunce86 — APPROVED** [review]
just nits

**Review by sunce86 — CHANGES_REQUESTED** [review]
Blocking the merge until solver team is ready to support multiple fees. 

The alternative is to add configuration parameter to allow multiple fees and if enabled, combine both partner fee and default autopilot fee policy, otherwise attach only partner fee to order.

Let me know what you think. @MartinquaXD I am ok with both, maybe slightly in favor of adding config so we can merge the PR.

**Comment by MartinquaXD** [comment]
> maybe slightly in favor of adding config so we can merge the PR.

Yeah, I think since the solver team asked to delay enabling this adding a config parameter would be slightly nicer to get this code into `main` soon.

**Comment by Mateo-mro** [comment]
I added a configuration argument for the autopilot to enable the multiple fees (disabled by default).
cc @sunce86

**Review by MartinquaXD — APPROVED** [review]

File: crates/autopilot/src/domain/fee/mod.rs [hunk 25]
```diff
@@ -122,28 +126,83 @@ impl ProtocolFees {
...
+        quote_: boundary::Amounts,
+        partner_fees: Vec<Policy>,
+    ) -> domain::Order {
+        if let Some(partner_fee) = partner_fees.first() {
+            return boundary::order::to_domain(order, vec![*partner_fee]);
+        }
         let protocol_fees = self
             .fee_policies
             .iter()
-            .filter_map(|fee_policy| {
```
[the remark this answers is not in the document]
Comment (Mateo-mro) [reply crates/autopilot/src/domain/fee/mod.rs]: Yeah, that was exactly my thought too, that's why I didn't want to invest much more time in making it cleaner. I refactor common code in functions, so at least "isn't that bad".

**Review by sunce86 — APPROVED** [review]
Lg
